### PR TITLE
Adding widgets for largest incidents and largest inputs and outputs

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/26523

## Description
Adding widgets for largest incidents and largest inputs and outputs

## Screenshots
image

### CODE REVIEW

File: Packs/CommonWidgets/Scripts/LargestInputsAndOuputsInIncidents/LargestInputsAndOuputsInIncidents.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        'data': sorted(largest_inputs_and_outputs, key=itemgetter('Size'), reverse=True),
+        'total': len(largest_inputs_and_outputs)
+    }
+
+    return widget_table
+
+
+def main():
+    try:
+        raw_output = demisto.executeCommand('LargestInvestigationsWidget', args={})
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/LargestInputsAndOuputsInIncidents/LargestInputsAndOuputsInIncidents.py:66]: you will need to update the args here once you add the fromDate

**Comment by ronykoz** [comment]
@meirwah please have a look at the widget images posted above

**Comment by meirwah** [comment]
LGTM.
but naming of both could be better, plz consult UX on that

**Comment by bakatzir** [comment]
This PR includes the 2 scripts, should we include OOTB the widget itself as well at `Packs/CommonWidgets/Widgets`?

**Comment by ronykoz** [comment]
@bakatzir thats the plan

**Comment by meirwah** [comment]
Is this PR still been worked on? where is it blocked?

File: Packs/CommonWidgets/Scripts/LargestInputsAndOuputsInIncidents/LargestInputsAndOuputsInIncidents.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+comment: Create widget for the largest investigations
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/LargestInputsAndOuputsInIncidents/LargestInputsAndOuputsInIncidents.yml:14]: done

File: Packs/CommonWidgets/Scripts/LargestInvestigationsWidget/LargestInvestigationsWidget.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    return investigations
+
+
+def parse_investigations_to_table(investigations):
+    data: List = []
+    widget_table = {"total": len(investigations)}
+    for investigation in investigations.keys():
+        size = investigations[investigation].get('leafSize').split(' ')
+        if float(size[0]) >= 1.0 and size[1] == 'MB':
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/LargestInvestigationsWidget/LargestInvestigationsWidget.py:26]: yes, as @meirwah stated in the issue this is what we are looking for

File: Packs/CommonWidgets/Scripts/LargestInvestigationsWidget/LargestInvestigationsWidget.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+comment: Create widget for the largest investigations
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/LargestInvestigationsWidget/LargestInvestigationsWidget.yml:14]: done

**Comment by ronykoz** [comment]
@meirwah I think it's in its in the final steps, @roysagi is adding UT now.

File: Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
+{
+ "id": "LargestInputsAndOuputsInIncidentsWidget",
+ "version": -1,
+ "modified": "2020-08-18T13:31:21.939387+03:00",
+ "name": "Largest Inputs And Ouputs In Incidents",
+ "dataType": "scripts",
+ "widgetType": "table",
+ "query": "GetLargestInputsAndOuputsInIncidents",
+ "dateRange": {
+  "fromDate": "0001-01-01T00:00:00Z",
```
Comment (ronykoz) [inline Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json:10]: should this be filled?

File: Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
+{
+ "id": "LargestInvestigationsWidget",
+ "version": -1,
+ "modified": "2020-08-18T13:28:59.339259+03:00",
+ "name": "Largest Investigations",
+ "dataType": "scripts",
+ "widgetType": "table",
+ "query": "GetLargestInvestigationst",
+ "dateRange": {
+  "fromDate": "0001-01-01T00:00:00Z",
```
Comment (ronykoz) [inline Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json:10]: same here?

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+- default: false
+  description: The date whose month is the last month we want to fetch the incidents.
+  isArray: false
+  name: to
+  required: false
+  secret: false
+- default: false
+  description: The date whose month is the first month we want to fetch the incidents.
+  isArray: false
+  name: from
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml:11]: shouldn't this be fromDate?
Same goes for the other yml

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml [hunk 8]
```diff
@@ -0,0 +1,27 @@
+args:
+- default: false
+  description: The date whose month is the last month we want to fetch the incidents.
+  isArray: false
+  name: to
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml:5]: same here

File: Packs/CommonWidgets/Scripts/GetLargestInputsAndOuputsInIncidents/GetLargestInputsAndOuputsInIncidents_test.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    'IncidentID': 1,
+    'TaskID': '200',
+    'TaskName': 'Malware Investigation 2',
+    'Name': 'IP',
+    'Size': 200.692,
+    'InputOrOutput': 'Output'
+}
+
+
+def test_get_largest_inputs_and_outputs():
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/GetLargestInputsAndOuputsInIncidents/GetLargestInputsAndOuputsInIncidents_test.py:45]: let's add when given then

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            })
+
+    widget_table['data'] = sorted(data, key=itemgetter('Size'), reverse=True)  # type: ignore
+
+    return widget_table
+
+
+def get_month_db_from_date(date):
+    month = date.strftime('%m')
+    year = date.strftime('%Y')
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.py:42]: why only in months?

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations_test.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    'investigations-4187': {
+        'keyN': 1301,
+        'leafSize': '4.8 MB',
+        'Date': '082020'
+    }
+}
+
+
+def test_get_investigations():
```
Comment (ronykoz) [inline Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations_test.py:43]: add when given then

**Review by ronykoz — CHANGES_REQUESTED** [review]

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.py (hunk 10 above)
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.py:42]: you can only filter the `getDBStatistics` command by db names which are effectively months (for example 082020). There are no further date-time info about the incidents from the output so this is the best I can do for time control.

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml (hunk 7 above)
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml:11]: no, checked this it is definitely `from` and `to`

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml (hunk 8 above)
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml:5]: same as below

File: Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json (hunk 5 above)
Comment (roysagi) [reply Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json:10]: It was set up like that via `sdk format` and it seems other widgets are built the same way

File: Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json (hunk 6 above)
Comment (roysagi) [reply Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json:10]: same as above

File: Packs/CommonWidgets/Scripts/GetLargestInputsAndOuputsInIncidents/GetLargestInputsAndOuputsInIncidents_test.py (hunk 9 above)
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/GetLargestInputsAndOuputsInIncidents/GetLargestInputsAndOuputsInIncidents_test.py:45]: done

File: Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations_test.py (hunk 11 above)
Comment (roysagi) [reply Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations_test.py:43]: done

**Review by ronykoz — APPROVED** [review]

**Comment by roysagi** [comment]
@kirbles19 - if you could please check the ymls for my two scripts:
1) [here](https://github.com/demisto/content/blob/widgets_big_outputs/Packs/CommonWidgets/Scripts/GetLargestInputsAndOuputsInIncidents/GetLargestInputsAndOuputsInIncidents.yml)
2) and [here](https://github.com/demisto/content/blob/widgets_big_outputs/Packs/CommonWidgets/Scripts/GetLargestInvestigations/GetLargestInvestigations.yml)

**Comment by kirbles19** [comment]
@roysagi Done.

**Comment by roysagi** [comment]
@ronykoz - validation will keep failing until next SDK release when this https://github.com/demisto/demisto-sdk/pull/680 gets in

File: Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    "key": "TaskID"
+   },
+   {
+    "displayed": true,
+    "isDefault": true,
+    "key": "TaskName"
+   }
+  ]
+ },
+ "fromVersion": "5.0.0",
```
Comment (meirwah) [inline Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json:56]: why is this from 5.0?
this feature is for 6.0 and above only

File: Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    "key": "Date"
+   },
+   {
+    "displayed": true,
+    "isDefault": true,
+    "key": "AmountOfEntries"
+   }
+  ]
+ },
+ "fromVersion": "5.0.0",
```
Comment (meirwah) [inline Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json:46]: why is this from 5.0?
this feature is for 6.0 and above only

Comment (roysagi) [reply Packs/CommonWidgets/Widgets/widget-LargestInvestigationsWidget.json]: done

File: Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json (hunk 12 above)
Comment (roysagi) [reply Packs/CommonWidgets/Widgets/widget-LargestInputsAndOuputsInIncidentsWidget.json]: done

**Comment by meirwah** [comment]
what is status here? 
what's the ETA for this to be in?

**Comment by ronykoz** [comment]
@meirwah waiting for a new version of demisto-sdk, should be out on Wednesday

**Pull request merged**